hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterServiceImpl.java

```diff
@@ -604,15 +604,17 @@ public void handleJoinRequest(JoinRequest joinRequest, Connection connection) {
             return;
         }
 
-        if (!node.isMaster()) {
-            sendMasterAnswer(joinRequest.getAddress());
+        Address target = joinRequest.getAddress();
+        boolean isRequestFromCurrentMaster = target.equals(node.getMasterAddress());
```

Comment:
Looks good. Nice catch. Can you add some explanation about the reason of this check as a comment? I don't want to miss the logic here.